enterprise/kerberos-login-module/src/main/java/com.hazelcast.security/KerberosLoginModule.java

```diff
@@ -0,0 +1,196 @@
+        }
+
+        ExtendedGSSContext extendedContext = (ExtendedGSSContext) gssContext;
+
+        try {
+            Object authzDataObject = extendedContext.inquireSecContext(InquireType.KRB5_GET_AUTHZ_DATA);
+            if (authzDataObject == null) {
+                throw new FailedLoginException("Internal error (no authz data).");
+            }
+
+            com.sun.security.jgss.AuthorizationDataEntry[] authzEntries = (com.sun.security.jgss.AuthorizationDataEntry[]) authzDataObject;
+            com.sun.security.jgss.AuthorizationDataEntry authzEntry = null;
+            sun.security.krb5.internal.AuthorizationDataEntry pacAuthzEntry = null;
+            for (int i = 0; i < authzEntries.length; i++) {
+                if (authzEntries[i].getType() != AD_IF_RELEVANT) continue;
+                authzEntry = authzEntries[i];
+
+                byte[] adData = authzEntries[i].getData();
+                DerInputStream adDataStream = new DerInputStream(adData);
+                DerValue[] values = adDataStream.getSet(adData.length, true);
+                DerValue value = values[0]; // should be the first one
+
+                try {
+                    pacAuthzEntry = new sun.security.krb5.internal.AuthorizationDataEntry(value);
+                    if (pacAuthzEntry.adType == AD_WIN2K_PAC) break;
+                }
+                catch (IOException e) {
+                    // swallow
+                }
+
+                pacAuthzEntry = null;
+            }
+
+            if (authzEntry == null) {
+                throw new FailedLoginException("Internal error (no AD_IF_RELEVANT entry).");
+            }
+
+            if (pacAuthzEntry == null) {
+                throw new FailedLoginException("Internal error (no AD-WIN2K-PAC entry within the AD-IF-RELEVANT entry).");
+            }
+
+            // add all groups as roles
+            // note: these are SIDs, not group names - we don't have the names in the Kerberos ticket
+            try {
+                RpcSid[] sids = new Pac(pacAuthzEntry.adData).getGroupMemberships();
+                for (int i = 0; i < sids.length; i++) {
+                    //logger.info("sid[" + i + "]: " + sids[i].toString());
+                    addRole(StringUtil.lowerCaseInternal(sids[i].toString()));
+                }
+            }
+            catch (Exception e) {
+                throw new FailedLoginException("Failed to register roles. " + e);
+            }
+        }
+        catch (GSSException | IOException | Asn1Exception e) {
+            throw new FailedLoginException("Failed to validate access. " + e.getMessage());
+        }
+    }
+
+    public boolean onLogin() throws LoginException {
+
+        // get the token
+        byte[] token = getToken();
+
+        // accept the token
+        GSSContext gssContext = acceptToken(token);
+
+        // validate access (throws if access is denied)
+        validateAccess(gssContext);
```

Comment:
I would rename the method to a more descriptive name here and wrap it into the `skipRole` condition.
E.g.

```java
if (! isSkipRole()) {
  addGroupSidsFromMsPac(token);
}
```

Reply:
Understood. Refactoring.